README.md
OWASP Bricks Cookbook
=============
Deploy a bricks environment.
[![Cookbook Version](https://img.shields.io/cookbook/v/bricks.svg)](https://community.opscode.com/cookbooks/bricks) [![Build Status](https://secure.travis-ci.org/wargames-cookbooks/bricks.png)](http://travis-ci.org/wargames-cookbooks/bricks)

Requirements
------------

#### Platform
- `Ubuntu 14.04`
- `Debian 8.7`

#### Cookbooks
- `apache2` - https://supermarket.chef.io/cookbooks/apache2
- `mysql` - https://supermarket.chef.io/cookbooks/mysql
- `mysql2_chef_gem` - https://supermarket.chef.io/cookbooks/mysql2_chef_gem
- `php` - https://supermarket.chef.io/cookbooks/php
- `database` - https://supermarket.chef.io/cookbooks/database

#### Supported versions - codename
- 2.2 - Tuivai
- 2.1 - Mora
- 2.0 - Dakatua
- 1.9 - Barak
- 1.8 - Atrai
- 1.7 - Phalgu


Attributes
----------
#### bricks::default
| Key                               | Type   |  Description                                             |
| ------------------------ | ------- | ---------------------------------------------------------------- |
| `[bricks][db][name]`     | String  | Database name (default: `bricks`)                                |
| `[bricks][db][username]` | String  | Database user name (default: `bricks`)                           |
| `[bricks][db][password]` | String  | Database user password (default: `bricks`)                       |
| `[bricks][codename]`     | String  | Bricks codename to deploy (default: `:tuivai`)                   |
| `[bricks][path]`         | String  | Path where application will be deployed (default: `/opt/bricks`) |
| `[bricks][showhint]`     | Boolean | DVWA default security level (default: `false`)                   |
| `[bricks][server_name]`  | String  | Apache2 server name (default: `bricks`)                          |
| `[bricks][server_name]`  | Array   | Array of apache2 virtualhost aliases (default: `[bricks]`)       |

Usage
-----
#### bricks::default

Just include `bricks` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[bricks]"
  ]
}
```

#### Running tests

- First, install dependencies:  
`bundle install`

- Run Checkstyle and ChefSpec:  
`bundle exec rake`

- Run Kitchen tests:  
`bundle exec rake kitchen`  

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add-component-x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: Sliim <sliim@mailoo.org> 

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
# CHANGELOG for Bricks

This file is used to list changes made in each version of bricks cookbook.

## 0.3.0:

* Chef 13 compatibility
* Updated supported platforms:
  - Ubuntu 14.04
  - Debian 8.7

## 0.2.1:

* Install mysql2 gem package with `mysql2_chef_gem` cookbook

## 0.2.0:

* Chef 12 compatibility
* Use Rake instead of Strainer
* Test harness (rubocop, chefspec, kitchen)
* New Brick's codename support (tuivai, mora, dakatua)
* Debian 7 support

## 0.1.4:

* Changed license for Apache 2.0

## 0.1.3:

* Integration testing with serverspec and test-kitchen

## 0.1.2:

* Added Atrai and Barak codename support.

## 0.1.1:

* Added Phalgu codename support.
* Added a bit of readme.

## 0.1.0:

* Initial release of bricks


spec/default_spec.rb
# -*- coding: utf-8 -*-

require 'rspec/expectations'
require 'chefspec'
require 'chefspec/berkshelf'

ChefSpec::Coverage.start! { add_filter 'bricks' }

require 'chef/application'

describe 'bricks::default' do
  before do
    stub_command('/usr/sbin/apache2 -t').and_return(true)
  end

  let(:subject) do
    ChefSpec::SoloRunner.new(file_cache_path: '/var/chef/cache',
                             platform: 'debian',
                             version: '9.0') do |node|
      node.override['apache']['mpm'] = 'prefork'
      node.override['bricks']['codename'] = :phalgu
      node.override['bricks']['server_name'] = 'bricks-server'
      node.override['bricks']['showhint'] = true
      node.override['bricks']['path'] = '/opt/bricks-app'
      node.override['bricks']['db']['username'] = 'bricksuser'
      node.override['bricks']['db']['password'] = 'brickspass'
      node.override['bricks']['db']['name'] = 'bricksdb'
    end.converge(described_recipe)
  end

  it 'should install unzip package' do
    expect(subject).to install_package('unzip')
  end

  it 'should include required recipes for webapp' do
    expect(subject).to include_recipe('apache2')
    expect(subject).to include_recipe('apache2::mod_php')
    expect(subject).to include_recipe('php')
    expect(subject).to include_recipe('php::module_mysql')
  end

  it 'should download bricks zip file' do
    expect(subject).to create_remote_file('/var/chef/cache/bricks.zip')
      .with(source: 'http://downloads.sourceforge.net/project/owaspbricks/Phalgu%20-%201.7/OWASP%20Bricks%20-%20Phalgu.zip')
  end

  it 'should create bricks directory' do
    expect(subject).to create_directory('/opt/bricks-app')
  end

  it 'should unzip bricks archive' do
    expect(subject).to run_execute('unzip-bricks')
      .with(cwd: '/var/chef/cache',
            command: 'unzip -u bricks.zip')
  end

  it 'should copy bricks source code in bricks rootdir' do
    expect(subject).to run_execute('copy-bricks-source')
      .with(cwd: '/var/chef/cache',
            command: 'cp -rf bricks/* /opt/bricks-app')
  end

  it 'should create config file from template' do
    config_file = '/opt/bricks-app/LocalSettings.php'
    matches = [/^\$dbuser\s+=\s+'bricksuser';$/,
               /^\$dbpass\s+=\s+'brickspass';$/,
               /^\$dbname\s+=\s+'bricksdb';$/,
               /^\$host\s+=\s+'localhost';$/,
               /^\$showhint\s+=\s+true;$/,
               %r{^\$server\s+=\s+'http:\/\/bricks-server';$}]

    expect(subject).to create_template(config_file)
      .with(source: 'LocalSettings.php.erb',
            owner: 'www-data',
            group: 'www-data')

    matches.each do |m|
      expect(subject).to render_file(config_file).with_content(m)
    end
  end

  it 'should install mysql2 gem package' do
    expect(subject).to install_mysql2_chef_gem('default')
  end

  it 'should setup mysql service' do
    expect(subject).to create_mysql_service('default')
      .with(port: '3306',
            version: '5.5',
            initial_root_password: 'toor')
  end

  it 'should drop mysql database' do
    expect(subject).to drop_mysql_database('drop-bricks-db')
      .with(database_name: 'bricksdb',
            connection: { host: '127.0.0.1',
                          username: 'root',
                          password: 'toor',
                          socket: '/run/mysql-default/mysqld.sock' })
  end

  it 'should create mysql database' do
    expect(subject).to create_mysql_database('create-bricks-db')
      .with(database_name: 'bricksdb',
            connection: { host: '127.0.0.1',
                          username: 'root',
                          password: 'toor',
                          socket: '/run/mysql-default/mysqld.sock' })
  end

  it 'should create database user' do
    expect(subject).to grant_mysql_database_user('bricksuser')
      .with(password: 'brickspass',
            database_name: 'bricksdb',
            privileges: [:select, :update, :insert, :create, :delete, :drop],
            connection: { host: '127.0.0.1',
                          username: 'root',
                          password: 'toor',
                          socket: '/run/mysql-default/mysqld.sock' })
  end

  it 'should populate bricks database with mysql dump' do
    expect(subject).to run_execute('import-mysql-dump')
      .with(command: 'mysql -h 127.0.0.1 -u root -ptoor '\
                     '--socket /run/mysql-default/mysqld.sock '\
                     'bricksdb < /opt/bricks-app/config/bricks.sql')
  end
end


test/cookbooks/bricks_test/metadata.rb
# This source file is part of Bricks's chef cookbook.
#
# Bricks's chef cookbook is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Bricks's chef cookbook is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Bricks's chef cookbook. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
#

name              "bricks_test"
maintainer        "Sliim"
maintainer_email  "sliim@mailoo.org"
license           "GPLv3"
description       "Test cookbook for bricks."
long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version           "0.1.0"


test/cookbooks/bricks_test/recipes/default.rb
# This source file is part of Bricks's chef cookbook.
#
# Bricks's chef cookbook is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Bricks's chef cookbook is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Bricks's chef cookbook. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
#
# Cookbook Name:: bricks_test
# Recipe:: default
#

include_recipe "bricks::default"


metadata.rb
# -*- coding: utf-8 -*-
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

name 'bricks'
maintainer 'Sliim'
maintainer_email 'sliim@mailoo.org'
license 'Apache-2.0'
description 'Installs/Configures Bricks application'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
chef_version '>= 12.5' if respond_to?(:chef_version)
version '0.3.0'

recipe 'default', 'Installs/configures bricks webapp + db'

depends 'apache2'
depends 'mysql'
depends 'mysql2_chef_gem'
depends 'php'
depends 'database'

supports 'ubuntu'
supports 'debian'

source_url 'https://github.com/wargames-cookbooks/bricks' if
  respond_to?(:source_url)
issues_url 'https://github.com/wargames-cookbooks/bricks/issues' if
  respond_to?(:issues_url)


recipes/default.rb
# -*- coding: utf-8 -*-
#
# Cookbook Name:: bricks
# Recipe:: default
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

codename_map = {
  tuivai: 'http://downloads.sourceforge.net/project/owaspbricks/Tuivai%20-%202.2/OWASP%20Bricks%20-%20Tuivai.zip',
  mora: 'http://downloads.sourceforge.net/project/owaspbricks/Mora%20-%202.1/OWASP%20Bricks%20-%20Mora.zip',
  dakatua: 'http://downloads.sourceforge.net/project/owaspbricks/Dakatua%20-%202.0/OWASP%20Bricks%20-%20Dakatua.zip',
  barak: 'http://downloads.sourceforge.net/project/owaspbricks/Barak%20-%201.9/OWASP%20Bricks%20-%20Barak.zip',
  atrai: 'http://downloads.sourceforge.net/project/owaspbricks/Atrai%20-%201.8/OWASP%20Bricks%20-%20Atrai.zip',
  phalgu: 'http://downloads.sourceforge.net/project/owaspbricks/Phalgu%20-%201.7/OWASP%20Bricks%20-%20Phalgu.zip',
}

unless codename_map.key?(node['bricks']['codename'].to_sym)
  raise ArgumentError, "Invalid bricks codename: #{node['bricks']['codename']}"
end

package 'unzip'

include_recipe 'apache2'
include_recipe 'apache2::mod_php'
include_recipe 'php'
include_recipe 'php::module_mysql'

remote_file "#{Chef::Config[:file_cache_path]}/bricks.zip" do
  source codename_map[node['bricks']['codename'].to_sym]
end

directory node['bricks']['path'] do
  recursive true
end

execute 'unzip-bricks' do
  cwd Chef::Config[:file_cache_path]
  command 'unzip -u bricks.zip'
end

execute 'copy-bricks-source' do
  cwd Chef::Config[:file_cache_path]
  command "cp -rf bricks/* #{node['bricks']['path']}"
end

template "#{node['bricks']['path']}/LocalSettings.php" do
  source 'LocalSettings.php.erb'
  owner 'www-data'
  group 'www-data'
  variables db: node['bricks']['db'],
            server: "http://#{node['bricks']['server_name']}",
            host: 'localhost',
            showhint: node['bricks']['showhint'].to_s
end

# Apache2 configuration
apache_site 'default' do
  enable false
end

web_app 'bricks' do
  enable true
  docroot node['bricks']['path']
  server_name node['bricks']['server_name']
  server_aliases node['bricks']['server_aliases']
end

# MySQL configuration
connection_info = {
  host: '127.0.0.1',
  username: 'root',
  password: 'toor',
  socket: '/run/mysql-default/mysqld.sock',
}

mysql2_chef_gem 'default' do
  package_version '5.5'
end

mysql_service 'default' do
  port '3306'
  version '5.5'
  initial_root_password 'toor'
  action [:create, :start]
end

mysql_database 'drop-bricks-db' do
  connection connection_info
  database_name node['bricks']['db']['name']
  action :drop
end

mysql_database 'create-bricks-db' do
  connection connection_info
  database_name node['bricks']['db']['name']
end

mysql_database_user node['bricks']['db']['username'] do
  connection connection_info
  password node['bricks']['db']['password']
  database_name node['bricks']['db']['name']
  privileges [:select, :update, :insert, :create, :delete, :drop]
  action :grant
end

execute 'import-mysql-dump' do
  command "mysql -h #{connection_info[:host]} "\
          "-u #{connection_info[:username]} "\
          "-p#{connection_info[:password]} "\
          '--socket /run/mysql-default/mysqld.sock '\
          "#{node['bricks']['db']['name']} "\
          "< #{node['bricks']['path']}/config/bricks.sql"
end
